tests/Aql/IndexNodeTest.cpp

```diff
@@ -0,0 +1,982 @@
+                                        std::string const& optionsString = "{}"
+) {
+  arangodb::aql::Query query(false, vocbase, arangodb::aql::QueryString(queryString), bindVars,
+                             arangodb::velocypack::Parser::fromJson(optionsString),
+                             arangodb::aql::PART_MAIN);
+
+  std::shared_ptr<arangodb::aql::SharedQueryState> ss = query.sharedState();
+
+  arangodb::aql::QueryResult result;
+  while (true) {
+    auto state = query.execute(arangodb::QueryRegistryFeature::registry(), result);
+    if (state == arangodb::aql::ExecutionState::WAITING) {
+      ss->waitForAsyncResponse();
+    } else {
+      break;
+    }
+  }
+  return result;
+}
+
+TEST_F(IIndexNodeTest, objectQuery) {
+  TRI_vocbase_t vocbase(TRI_vocbase_type_e::TRI_VOCBASE_TYPE_NORMAL, createInfo(server.server()));
+  // create a collection
+  auto collectionJson = arangodb::velocypack::Parser::fromJson(
+    "{\"name\": \"testCollection\", \"id\": 42}");
+  auto collection = vocbase.createCollection(collectionJson->slice());
+  ASSERT_FALSE(!collection);
+  auto indexJson = arangodb::velocypack::Parser::fromJson("{\"type\": \"hash\", \"fields\": [\"obj.a\", \"obj.b\", \"obj.c\"]}");
+  auto createdIndex = false;
+  auto index = collection->createIndex(indexJson->slice(), createdIndex);
+  ASSERT_TRUE(createdIndex);
+  ASSERT_FALSE(!index);
+
+  std::vector<std::string> const EMPTY;
+  arangodb::transaction::Methods trx(arangodb::transaction::StandaloneContext::Create(vocbase),
+                                     EMPTY, EMPTY, EMPTY,
+                                     arangodb::transaction::Options());
+  EXPECT_TRUE(trx.begin().ok());
+
+  arangodb::OperationOptions opt;
+  arangodb::ManagedDocumentResult mmdoc;
+  auto jsonDocument = arangodb::velocypack::Parser::fromJson("{\"obj\": {\"a\": \"a_val\", \"b\": \"b_val\", \"c\": \"c_val\"}}");
+  auto const res = collection->insert(&trx, jsonDocument->slice(), mmdoc, opt, false);
+  EXPECT_TRUE(res.ok());
+  EXPECT_TRUE(trx.commit().ok());
+
+  {
+    auto queryString = "FOR d IN testCollection FILTER d.obj.a == 'a_val' SORT d.obj.c LIMIT 10 RETURN d";
+    auto queryResult = ::executeQuery(vocbase, queryString);
+    EXPECT_TRUE(queryResult.result.ok()); // commit
+    auto result = queryResult.data->slice();
+    EXPECT_TRUE(result.isArray());
+    arangodb::velocypack::ArrayIterator resultIt(result);
+    ASSERT_EQ(1, resultIt.size());
+  }
+
+  // const object in condition
+  {
+    auto queryString = "FOR d IN testCollection FILTER d.obj.a == {sub_a: \"a_val\"}.sub_a SORT d.obj.c LIMIT 10 RETURN d";
+    auto queryResult = ::executeQuery(vocbase, queryString);
+    EXPECT_TRUE(queryResult.result.ok()); // commit
+    auto result = queryResult.data->slice();
+    EXPECT_TRUE(result.isArray());
+    arangodb::velocypack::ArrayIterator resultIt(result);
+    ASSERT_EQ(1, resultIt.size());
```

Comment:
Should we check actual returned document, just to be sure all worked as expected?